pkg/services/accesscontrol/resourcepermissions/api_adapter.go

```diff
@@ -162,3 +165,223 @@ func getMapKeys(m map[string][]string) []string {
+
+	existing, err := resourcePermResource.Get(ctx, resourcePermName, metav1.GetOptions{})
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get existing resource permission: %w", err)
+	}
+
+	var existingPermissions []interface{}
+	if existing != nil {
+		existingPermissions, _, _ = unstructured.NestedSlice(existing.Object, "spec", "permissions")
+	}
+
+	newPermissions := make([]interface{}, 0)
+
+	for _, permRaw := range existingPermissions {
+		permMap, ok := permRaw.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		existingKind, _, _ := unstructured.NestedString(permMap, "kind")
+		existingName, _, _ := unstructured.NestedString(permMap, "name")
+
+		if existingKind == kind && existingName == name {
+			continue
+		}
+
+		newPermissions = append(newPermissions, permMap)
+	}
+
+	if permission != "" {
+		newPermissions = append(newPermissions, map[string]interface{}{
+			"kind": kind,
+			"name": name,
+			"verb": permission,
+		})
+	}
+
+	if len(newPermissions) == 0 {
+		if existing != nil {
+			err = resourcePermResource.Delete(ctx, resourcePermName, metav1.DeleteOptions{})
+			if err != nil && !k8serrors.IsNotFound(err) {
+				return fmt.Errorf("failed to delete resource permission in k8s: %w", err)
+			}
+		}
+		return nil
+	}
+
+	resourcePerm := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": iamv0.ResourcePermissionInfo.GroupVersion().String(),
+			"kind":       iamv0.ResourcePermissionInfo.TypeMeta().Kind,
+			"metadata": map[string]interface{}{
+				"name":      resourcePermName,
+				"namespace": namespace,
+			},
+			"spec": map[string]interface{}{
+				"resource": map[string]interface{}{
+					"apiGroup": a.getAPIGroup(),
+					"resource": a.service.options.Resource,
+					"name":     resourceID,
+				},
+				"permissions": newPermissions,
+			},
+		},
+	}
+
+	if existing != nil {
+		resourcePerm.SetResourceVersion(existing.GetResourceVersion())
+		_, err = resourcePermResource.Update(ctx, resourcePerm, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to update resource permission in k8s: %w", err)
+		}
+	} else if permission != "" {
+		_, err = resourcePermResource.Create(ctx, resourcePerm, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to create resource permission in k8s: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (a *api) getPermissionKind(perm accesscontrol.SetResourcePermissionCommand) string {
+	if perm.UserID != 0 {
+		return string(iamv0.ResourcePermissionSpecPermissionKindUser)
+	}
+	if perm.TeamID != 0 {
+		return string(iamv0.ResourcePermissionSpecPermissionKindTeam)
+	}
+	if perm.BuiltinRole != "" {
+		return string(iamv0.ResourcePermissionSpecPermissionKindBasicRole)
+	}
+	return ""
+}
+
+func (a *api) getPermissionName(ctx context.Context, perm accesscontrol.SetResourcePermissionCommand) string {
```

Comment:
We should return an error and error out if we don't get the UID
```suggestion
func (a *api) getPermissionName(ctx context.Context, perm accesscontrol.SetResourcePermissionCommand) (string, error) {
```